include/lsst/afw/cameraGeom/Amplifier.h

```diff
@@ -101,188 +95,268 @@ class Amplifier {
      *  assembled post-ISR image: final xy0 - pre-raw xy0.
      */
-    lsst::geom::Extent2I getRawXYOffset() const { return _rawXYOffset; }
-    void setRawXYOffset(lsst::geom::Extent2I const &xy) { _rawXYOffset = xy; }
-    //@}
+    lsst::geom::Extent2I getRawXYOffset() const { return getFields().rawXYOffset; }
 
-    //@{
     /**
      * The bounding box of usable horizontal overscan pixels in the assembled,
      * untrimmed raw image.
      */
-    lsst::geom::Box2I getRawHorizontalOverscanBBox() const { return _rawHorizontalOverscanBBox; }
-    void setRawHorizontalOverscanBBox(lsst::geom::Box2I const &bbox) { _rawHorizontalOverscanBBox = bbox; }
-    //@}
+    lsst::geom::Box2I getRawHorizontalOverscanBBox() const { return getFields().rawHorizontalOverscanBBox; }
 
-    //@{
     /**
      * The bounding box of usable vertical overscan pixels in the assembled,
      * untrimmed raw image.
      */
-    lsst::geom::Box2I getRawVerticalOverscanBBox() const { return _rawVerticalOverscanBBox; }
-    void setRawVerticalOverscanBBox(lsst::geom::Box2I const &bbox) { _rawVerticalOverscanBBox = bbox; }
-    //@}
+    lsst::geom::Box2I getRawVerticalOverscanBBox() const { return getFields().rawVerticalOverscanBBox; }
 
-    //@{
     /**
      * The bounding box of usable (horizontal) prescan pixels in the assembled,
      * untrimmed raw image.
      */
-    lsst::geom::Box2I getRawPrescanBBox() const { return _rawPrescanBBox; }
-    void setRawPrescanBBox(lsst::geom::Box2I const &bbox) { _rawPrescanBBox = bbox; }
-    //@}
+    lsst::geom::Box2I getRawPrescanBBox() const { return getFields().rawPrescanBBox; }
+
+protected:
+
+    struct Fields {
+        std::string name;
+        lsst::geom::Box2I bbox;
+        double gain = 0.0;
+        double readNoise = 0.0;
+        double saturation = 0.0;
+        double suspectLevel = 0.0;
+        ReadoutCorner readoutCorner = ReadoutCorner::LL;
+        ndarray::Array<double const, 1, 1> linearityCoeffs;
+        std::string linearityType;
+        double linearityThreshold;
+        double linearityMaximum;
+        std::string linearityUnits;
+        lsst::geom::Box2I rawBBox;
+        lsst::geom::Box2I rawDataBBox;
+        bool rawFlipX = false;
+        bool rawFlipY = false;
+        lsst::geom::Extent2I rawXYOffset;
+        lsst::geom::Box2I rawHorizontalOverscanBBox;
+        lsst::geom::Box2I rawVerticalOverscanBBox;
+        lsst::geom::Box2I rawPrescanBBox;
+    };
+
+    Amplifier() = default;
+
+    Amplifier(Amplifier const &) = default;
+    Amplifier(Amplifier &&) = default;
+    Amplifier & operator=(Amplifier const &) = default;
+    Amplifier & operator=(Amplifier &&) = default;
+
+    virtual Fields const & getFields() const = 0;
+
+};
+
+/**
+ *  A mutable Amplifier subclass class that can be used to incrementally
+ *  construct or modify Amplifiers.
+ */
+class Amplifier::Builder final : public Amplifier {
```

Comment:
My point is that as far as I can tell you *can* put a mutable object inside an immutable `Detector` or `Camera`. The only safeguard is that the existing code calls `finish`, which is easy to forget (especially since in some cases you deliberately pass Builders around, and that code may get cargo-culted into situations where it's not appropriate).

Reply:
"Annoying but doable" is an accurate description.

Reply:
> My point is that as far as I can tell you can put a mutable object inside an immutable Detector or Camera. The only safeguard is that the existing code calls finish, which is easy to forget (especially since in some cases you deliberately pass Builders around, and that code may get cargo-culted into situations where it's not appropriate).

Can you elaborate?  You can't get an immutable `Detector` or `Camera` without calling `Camera::Builder::finish()` or `Detector::PartialRebuilder::finish()`.  All of their constructors are either non-public or deleted, and those are the only methods that return those types.

Reply:
Or by calling `Detector::Factory`. Any member of `Detector` can call that constructor, and the only way to stop any of them from passing in an `Amplifier::Builder` is to be very careful when adding new features to the class.

I'm not going to argue about this further. But I want it on the record that I think future code changes to `Camera` and `Detector` will break this design, and I would never had agreed to RFC-632 had I noticed the word "subclass" in the original description.

Reply:
Understood.  I'll admit that I'm disregarding some good-in-general principles here because I think the circumstances make that the least bad option.  That includes being pretty confident that C++ changes of the sort that would open up those loopholes just won't happen, for reasons unrelated to the code itself - they're more about the skills and experience of the development team.